### ISSUE TITLE
BackStack in NavigationService in Windows Store Apps

### ISSUE THREAD

**Issue by DarryStonem** [PrismLibrary/Prism#67]
Why in Windows Store Apps I can't delete (or view) my backstack items?
Maybe it'll become handly for many users (incluiding me).

**Comment by bartlannoeye** [comment]
Related to #26

**Comment by briannoyes** [comment]
I'm open to adding some functionality for this. If anyone wants to take it on now that there is a starting point there and submit pull request, that will help get it there sooner.

**Comment by veccie** [comment]
Now that #78 has been merged into master, wouldn't it make sense to expose BackStack and BackStackDepth from IFrameFacade in INavigationService as well?

If yes then I'm unsure if we also need to create a wrapper/facade for PageStackEntry in the BackStack to hide some of its properties?

Currently the BackStack and BackStackDepth are not accessible from anywhere. Exposing these properties will make it possible to determine where the current page is in the navigation structure. When for example navigating to a sub-page you can then use this information and select the correct navigation head item.

**Comment by briannoyes** [comment]
I'm interested to hear what kind of use cases we would be addressing with stuff like this. Even the removal of items from the back stack was something I have not encountered a need for. Sounds a little like you want to be able to use the linear back stack that the platform offers to be able to do a more hierarchical back stack with branches and stuff like that, but that implies that the code doing all this direct manipulation of the back stack has to maintain the semantics by using the back stack in "tricky" ways. I'm just wondering if a different kind of navigation controller service might be a better fit for those kinds of requirements?

**Comment by bartlannoeye** [comment]
In the past I've had the need to pop pages from the backstack, which is available now. But for more complex manipulations I would indeed opt to write your own navigation service and keep the backstack itself hidden from the frontend developer.

**Comment by veccie** [comment]
I'm only talking about reading the BackStack, not manipulating it:) So the exposed BackStack could just be an `IEnumerable<T>`

Right now I'm using a SplitView navigation which is based on Microsoft's sample, but changed to be MVVM. I like the concept of it because it automatically updates the menu according the current page when you navigate back.

Look at OnNavigatingToPage in this code: https://github.com/Microsoft/Windows-universal-samples/blob/master/Samples/XamlNavigation/cs/AppShell.xaml.cs (2/3 down the page)

I've made my own ListView control which then listens for navigation events through PubSubEvent. It then selects the correct navigation item automatically based on the current page, which is really nice when you for example navigate back.

Now I don't have to think about where I'm at in the navigation structure :)

**Comment by briannoyes** [comment]
I see. That makes sense. Just to throw it out there another way I could think of to address that is to have a base class on my ViewModels that raises a PubSubEvent to signal what page was just navigated to. But I agree that being to build this into a list control would be a cleaner and more reusable way to handle that. And if we put that in maybe you would be willing to contribute your list control for inclusion in Prism to nicely handle the SplitView nav.

Anyone else have any reservations in exposing the back stack for observation like this?

**Comment by veccie** [comment]
I would be willing to contribute my list control to Prism.

About the base ViewModel class you mention. Right now I'm just forwarding the App.OnNavigated(..) to a ViewController which then publishes the PubSubEvent. But this can of course be done in other ways depending on the app.

I've started working on a new branch with my navigation sample. Once I've cleaned it up a bit I'll upload it to my fork so you can see/try it in action.

I'll create a new issue once it's done and we can discuss it further.

**Comment by eklavyamirani** [comment]
I had a scenario where I needed to manipulate the navigation stack. I had some temporary intermediate pages (Authentication pages) that didn't need to have a history in the navigation stack. I ended up writing a wrapper around the default NavigationService and added an abstraction to remove the page from the navigation stack once I navigated to the temporary page (In addition to normal linear navigation). 

I agree that scenarios like this are limited, but end up in a lot of apps anyway. We could either have an abstraction that allows us to do this (Navigate Without Storing in Stack) or as a best practice, just extend the navigation service to handle these scenarios individually. Curious what the best approach here might be.

**Comment by ali-hk** [comment]
The methods that were added don't update the SystemNavigationManager.AppViewBackButtonVisibility property in the scenario where DeviceGestureService.UseTitleBarBackButton is true.
This is currently done in PrismApplication.cs:OnNavigated, but that code has no way of knowing when NavigationService.CanGoBack() has actually changed but not triggered a navigation (such as when one of these three methods is called).

**Comment by briannoyes** [comment]
I'll look into it. We could introduce a dependency between the DeviceGestureService and the NavigationService so that the DGS could handle the navigation events and check CanGoBack to set that appropriately.

**Comment by ali-hk** [comment]
It seems to me that this is a case where a pub/sub event would make the most sense. Adding a hard dependency on DGS seems unnecessary.
In other words, have the FrameNavigationService publish a NavigationStateChanged event. One of the subscribers would be DGS which would then handle updating the back button. Publishing this event would also provide app code with an opportunity to listen for it and do any special processing needed (such as dismissing a SplitView-based menu).
If that sounds reasonable I can make the change and submit a pull request.

**Comment by briannoyes** [comment]
I think I like that approach. My only hesitation is that Prism itself has never depended on PubSub Events for any of its logic. But I agree this would be a good place to start because of the looser coupling and because as you said it makes the event available for any other app code that needs it. Go ahead and submit a PR please.
Thanks
Brian

**Comment by briannoyes** [comment]
As far as I can tell, this is addressed with @ali-hk and @jorisdebock 's PR's that were merged a long time ago. Closing, re-open with details if you are still blocked by something about back stack access.
